src/hotspot/share/cds/unregisteredClasses.cpp

```diff
@@ -24,97 +24,68 @@
-  Handle ext_class_name = java_lang_String::externalize_classname(name, CHECK_NULL);
-
-  JavaValue result(T_OBJECT);
-  JavaCallArguments args(3);
-  args.set_receiver(classloader);
-  args.push_oop(ext_class_name);
-  args.push_oop(super_class);
-  args.push_oop(interfaces);
-  JavaCalls::call_virtual(&result,
-                          UnregisteredClassLoader_klass(),
-                          methodName,
-                          methodSignature,
-                          &args,
-                          CHECK_NULL);
-  assert(result.get_type() == T_OBJECT, "just checking");
-  oop obj = result.get_oop();
-  return InstanceKlass::cast(java_lang_Class::as_Klass(obj));
-}
-
-class UnregisteredClasses::ClassLoaderTable : public ResourceHashtable<
-  Symbol*, OopHandle,
-  137, // prime number
-  AnyObj::C_HEAP> {};
+  assert(!_unregistered_class_loader.is_empty(), "not initialized");
+  Handle classloader(THREAD, _unregistered_class_loader.resolve());
 
-static UnregisteredClasses::ClassLoaderTable* _classloader_table = nullptr;
+  oop obj;
+  { // Call CDS$UnregisteredClassLoader::load(String name, String source)
+    Symbol* methodName = SymbolTable::new_symbol("load");
+    Symbol* methodSignature = SymbolTable::new_symbol("(Ljava/lang/String;Ljava/lang/String;)Ljava/lang/Class;");
+    Handle ext_class_name = java_lang_String::externalize_classname(name, CHECK_NULL);
+    Handle path_string = java_lang_String::create_from_str(path, CHECK_NULL);
 
-Handle UnregisteredClasses::create_classloader(Symbol* path, TRAPS) {
-  ResourceMark rm(THREAD);
-  JavaValue result(T_OBJECT);
-  Handle path_string = java_lang_String::create_from_str(path->as_C_string(), CHECK_NH);
-  Handle classloader = JavaCalls::construct_new_instance(
-                           UnregisteredClassLoader_klass(),
-                           vmSymbols::string_void_signature(),
-                           path_string, CHECK_NH);
-  return classloader;
-}
-
-Handle UnregisteredClasses::get_classloader(Symbol* path, TRAPS) {
-  if (_classloader_table == nullptr) {
-    _classloader_table = new (mtClass)ClassLoaderTable();
-  }
-  OopHandle* classloader_ptr = _classloader_table->get(path);
-  if (classloader_ptr != nullptr) {
-    return Handle(THREAD, (*classloader_ptr).resolve());
-  } else {
-    Handle classloader = create_classloader(path, CHECK_NH);
-    _classloader_table->put(path, OopHandle(Universe::vm_global(), classloader()));
-    path->increment_refcount();
-    return classloader;
+    JavaValue result(T_OBJECT);
+    JavaCalls::call_virtual(&result,
+                            classloader,
+                            UnregisteredClassLoader_klass(),
+                            methodName,
+                            methodSignature,
+                            ext_class_name,
+                            path_string,
+                            CHECK_NULL);
+    assert(result.get_type() == T_OBJECT, "just checking");
+    obj = result.get_oop();
```

Comment:
There's no need to put the above code in its own scope. Maybe do the following instead?

```
return InstanceKlass::cast(java_lang_Class::as_Klass(result.get_oop()));
```